src/merkle/smt/full/mod.rs

```diff
@@ -263,6 +286,92 @@ impl SparseMerkleTree<SMT_DEPTH> for Smt {
         leaf.hash()
     }
 
+    fn hash_prospective_leaf(&self, key: &RpoDigest, value: &Word) -> RpoDigest {
+        // This function combines logic from SmtLeaf::insert() and SmtLeaf::hash() to determine what
+        // the hash of a leaf would be with the `(key, value)` pair inserted into it, without simply
+        // cloning the leaf which could be expensive for some leaves, and is easily avoidable when
+        // we can combine the insertion and hashing operations.
```

Comment:
I would prefer to re-use existing code as much as possible here (even if it requires some extra cloning). The assumption is that collisions at depth 64 (i.e., when we have `SmtLeaf::Multiple`) would be extremely unlikely - and so it is OK to have somewhat less efficient code for these cases.

But also, we can create a separate issue for now to come back and optimize this more later.

Reply:
Ah, I did not realize multi-leaves would be so rare! If cloning the leaf is acceptable then this gets a *lot* easier :slightly_smiling_face:.